Implementation of thenCallOriginalImplementation

avandierast opened the pull request:
The implementation of thenCallOriginalImplementation required a refactor of the way self is discarded.
Now, self is kept in the params and discarded only when we don't need it.

Linked to #46 #48

I've also corrected some tests where the objects was used instead of the mock.

avandierast commented:
I'm a `black` user so part of the diff is due to `black`.... If that's a problem, I may revert the formating (or do another PR that does only the formating first)

kaste commented:
Very promising.  I wouldn't mind if you split the commit.  The fix of the ellipsis test can be one commit;  the small additional test in `staticmethods_test` can be one separate commit as well.

You need to add tests for using `thenCallOriginalImplementation` stubbing classmethods and staticmethods.  Typically a chaining test case like `when(os.path).exists(__file__).thenReturn(False).thenCallOriginalImplementation()` too.

Also define what happens if you use it on a method/function not present on the original spec/object.

The patch is easy too read.  I do not like that we wildly walk the objects so deep, e.g. `invocation.mock.mocked_obj` or even `invocation.mock.mocked_obj.__dict__` . I wonder why you use `mocked_obj` here, in `Mock.get_original_method` and `RememberedProxyInvocation.__call__` we access `mock.spec` for the original method/implementation.  Ideally the original implementation could be found on `mock.original_methods[method_name]` but we often store `None`s here so that would need a different refactoring before.  

I think it would be better if we decide that we must cut the first argument earlier.  We actually cut the argument late as you do here but decide if we have to early.  For example it must be clear when we construct `AnswerSelector` that we must cut self.  (E.g. `AnswerSelector(invocation, cut_self=True)`.)  -- because the answer only depends on the used method name (`invocation.method_name`) and the mock (or `mock.spec`).  We can compute a `cut_first_argument` when we construct the `Invocation` probably.

That being said; usually I'm happy with *lots* of tests so we can try different factorings, different ways to write the same thing, safely and later.

avandierast commented:
Hello :)

Thanks for the detailed response.
The idea with `AnswerSelector(..., cut_first_argument=...)` seems great to simplify the code. The chained sub-objects wild walk and the access to __dict__ may become useless with an implementation like this. I'll try to implement that.

avandierast commented:
Ok, deciding to cut self a bit earlier doesn't change the way we can retrieve the original function.

I didn't see `.original_methods[]` and it may be the cleanest way to get the original but it is mostly used for the unstubbing and thus is not really reliable as you said.
We can't use `get_original_method` because it doesn't return the original implementation anymore during the stubbing... 

Trying to get the original method on mock.spec doesn't reduce the complexity of the code since it's literally the same object as mocked_obj most of the time and when it is not, we have to read mocked_obj to get what we need.

Finally, after lots of attempt, I've decided to do a slight refacto of Mock to separate the unstub mecanism from the original methods retrieval.

The new implementation seems really more clean ! Maybe there is stil way for improvement or usefull tests to check.
Let me know :)

If you prefer the refacto in another PR that's also possible.

It still remains the definition of what happens when the mocked attribute is not present.

Adrian

kaste commented:
This is a *lot* harder to read than the first version.  This will take a sec.  

General note as we need to stay green per commit:  you can always add tests before implementation but mark them as `xfail`.  We should put all tests for this feature in one file otherwise it gets confusing.

kaste reviewed (commented):
There is some good stuff in here, some things I dislike.

I especially like that 

```
    def thenCallOriginalImplementation(self):
        answer = self.invocation.mock.get_original_method(
            self.invocation.method_name
        )
```

only relies on the `original_method` we actually get the same way as always,
t.i. with a `self.spec.__dict__[method_name]` in the normal case.  

Idiomatically you would do:

```
# in the init phase because stubbing *mutates* self.spec often 
fn = getattr(self.spec, method_name)
# and then
if hasattr(fn, '__self__'):
    fn = discard_self(fn)   # or: fn = fn.__func__   
``` 

That is very readable -- if it has self, e.g. `fn` *is* bound, unbind it.  But
`getattr` is a side-effect we want to avoid; objects (`specs`) could intercept
this.

So I do prefer relying on the `__dict__.get` variant as you do here to get the `original_method`.  

Interestingly 

```
        elif self.discard_first_arg and inspect.ismethod(answer):
            # Self is already bound, we can discard it.
            answer = discard_self(answer)
```

could also be

```
        elif inspect.ismethod(answer):
            # fn is bound, take the unbound function.
            answer = answer.__func__
```

That is of interest because it seems that we don't need the `discard_self` function (and with it the `self.discard_first_arg` information) for this case.


Also, perhaps surprisingly

```
    def thenReturn(self, *return_values):
        for return_value in return_values:
            answer = functools.partial(return_, return_value)
            if self.discard_first_arg:
                answer = discard_self(answer)
            self.__then(answer)
        return self

    def thenRaise(self, *exceptions):
        for exception in exceptions:
            answer = functools.partial(raise_, exception)
            if self.discard_first_arg:
                answer = discard_self(answer)
            self.__then(answer)
        return self
```

don't need to change at all.  T.i. they don't need `discard_first_arg` and `discard_self` at all because `return_` and `raise_` already eat *all* arguments.


I don't like though how you compute `should_discard_self` now.  I don't even think I understand it.   

For example, the difference between

```
    def get_mocked_method(self, method_name):
        mocked_obj = self.mocked_obj
        try:
            # We need to get the method and not a function
            # and staticmethod are not returned correctly if we use `getattr`
            return mocked_obj.__dict__[method_name]
        except (AttributeError, KeyError):
            return None
```

and

```
    def get_original_method(self, method_name):
        if self.spec is None:
            return None

        try:
            return self.spec.__dict__[method_name]
        except (AttributeError, KeyError):
            return getattr(self.spec, method_name, None)
```

is too subtle.  (Typically `self.spec == self.mocked_obj` holds!)  

I think because the original code decided to throw away `self` on `(mocked_obj, new_mocked_method)`: 
```
            if (
                inspect.ismethod(new_mocked_method)
                or inspect.isclass(self.mocked_obj)
                and not isinstance(new_mocked_method, staticmethod)
            ):
                args = args[1:]
```

you invented `get_mocked_method`.  As I understand it, `get_mocked_method` returns `new_mocked_method`.  But then the conditional in `should_discard_self`
is subtlety different

```
        and (
            # we throw away the first argument, if it's either self or cls
            inspect.ismethod(mocked_method)
            or inspect.isclass(spec)  # <== spec instead of mocked_obj
            and not isinstance(mocked_method, staticmethod)
            and not inspect.isclass(mocked_method)  # <== added
        )
```

Why is that?  (Actually the conditional in `should_discard_self` reads as if
`mocked_method` *is* `original_method` here.  Because `not inspect.isclass(original_method)` makes `new_mocked_method` a `staticmethod` in the original `replace_method` code. The very strange `# TBC: Inner classes` clause over there.)

I think we should expose a `should_discard_self(self, method_name)` on `Mock()` that only relies on `self.mocked_obj` and `original_method`.

avandierast commented:
Hello :)

Ok, your comment made things even clearer ! 
You're right, we only need to potentialy discard self in thenAnswer. I've modified the code, it works well.

About `should_discard_self`, it needs the

kaste commented:
Ok, doing `should_discard_self` based on the `original_method` is in tricky territory, but this one works for me:

```
    def have_prepended_arg(self, method_name):
        """ Returns if the method will have a prepedend self/class argument on call."""
        try:
            original_method = self._original_methods[method_name]
        except KeyError:
            return False
        else:
            return (
                inspect.ismethod(original_method)
                or (
                    inspect.isclass(self.mocked_obj)
                    and not isinstance(original_method, staticmethod)
                    and not inspect.isclass(original_method)
                )
            )
```

`have_prepended_arg` is not the better name.  It should be "has" not "have" anyway?  Usually `skip_first_arg` which is the same as `discard_first_arg` you already use.  Or `eat_self`.  Btw, I think the reference to a `self` is okay here even for `classmethods` because Python uses `__self__` in both cases.  And that we users name the first argument `cls` is really conventional only as we can give it any name we want.

kaste commented on mockito/invocation.py:
```diff
@@ -36,11 +35,14 @@
 
-__tracebackhide__ = operator.methodcaller(
-    "errisinstance",
-    (InvocationError, verificationModule.VerificationError)
-)
+# __tracebackhide__ = operator.methodcaller(
+#     "errisinstance",
+#     (InvocationError, verificationModule.VerificationError)
+# )
```
that's a mistake

avandierast replied:
Sorry... I failed my `git add`...

kaste commented on mockito/mocking.py:
```diff
@@ -28,10 +28,10 @@
-    "errisinstance",
-    invocation.InvocationError
-)
+# __tracebackhide__ = operator.methodcaller(
+#     "errisinstance",
+#     invocation.InvocationError
+# )
```
mistake as above

kaste commented on mockito/invocation.py:
```diff
@@ -410,23 +416,59 @@ def raise_(exception, *a, **kw):
+            self.invocation.method_name
+        )
+        if not answer:
+            raise AnswerError(
+                "The function '%s' of '%s' doesn't have any original implementation."
+                 % (self.invocation.method_name, self.invocation.mock.mocked_obj)
+            )
```
Maybe 
```suggestion
            raise AnswerError(
                "'%s' has no original implementation for '%s'." %
                (self.invocation.mock.mocked_obj, self.invocation.method_name)
            )
```

kaste commented on tests/call_original_implem_test.py:
```diff
@@ -0,0 +1,71 @@
+    def testNoOriginal(self):
+        dog = mock()
+        answer_selector = when(dog).bark()
+        with pytest.raises(AnswerError):
+            answer_selector.thenCallOriginalImplementation()
```
Checking the actual error message is a +.  We also get a better feel of how they would look like for the end-user.

```suggestion
        with pytest.raises(AnswerError) as exc:
            answer_selector.thenCallOriginalImplementation()

        assert str(exc.value) == (
            "'<class 'mockito.mocking.mock.<locals>.Dummy'>' "
            "has no original implementation for 'bark'."
        )
```

avandierast commented:
> Btw, I think the reference to a self is okay here even for classmethods because Python uses __self__ in both cases.
Ok, I changed because of classmethod :)

avandierast commented:
Ok, your implementation with the orignal_method works well. I've added a test on non-strict module function stubbing.

But it assumes that if original_method is None, then the new method can only be in two categories: module function and classic method. That's true because the `when` interface don't let us do anything else. I've added a comment about that because it seems usefull to not let that implicit.

Adrian

kaste commented:
I think this is okay.  Honestly I tried to implement this before but failed. 

Unfortunately there is a huge "But...".  The commits are not readable or tell a story.  You also merged master back in which makes `git rebase` a bit harder.  I cannot just squash because there are unrelated things in this PR.  I cannot simply cherry-pick as the single commits mix different matters as well.   (For example b191137 is a fixup for various other commits meshed into one.)  

I think what you should have done is push `fixup!` or `squash!` commits so there would be no need to `push --force` multiple times and we could apply them now to get better commits.  

1. It looks like you mangle the import section when you touch a file.  I *do* like the import style of your tool/script, or you do that manually?  But it is *not* a concern here.  Do this in one commit in a different PR or as the first commit I could cherry-pick.
2. Commit 5e8534a must be dropped.  We use `m = mock(); m.strict = False` here because we need a fake object as the first argument to `StubbedInvocation`.  There is no need to call it with a "real"  `Mock()` object, we're in the tests and want to fake the input here a bit.

If I see it correctly there are 6 commits you want to add to the project:

1. Restyle import section
2. Add a test to `tests/staticmethods_test.py`
3. Fix a typo in a method name in `tests/stubbing_test.py`
4. Add an (important!) test to `modulefunctions_test.py`
5. Rename `self.original_methods`-> `self._methods_to_unstub`
6. Implement `thenCallOriginalImplementation`

I probably should add comments inline.  

Tell me how your git skills are before *I* go down the partial cherry pick patch hole.

kaste commented on tests/modulefunctions_test.py:
```diff
@@ -105,3 +106,8 @@ def testUnstubFunctionOnModuleWhichIsActuallyAMethod_issue_53(self):
         unstub(random)
         assert random.randint(1, 10) != "mocked"
+
+    def testStubsNotStrict(self):
```
```suggestion
    def testAddFakeMethodInNotStrictMode(self):
```

The before untested functionality is to **add** new methods to an object.

kaste commented on tests/ellipsis_test.py:
```diff
@@ -373,23 +374,26 @@ def testKwargsMustBeUsedAsStarKwarg(self):
             when(rex).bark(*kwargs).thenReturn('Miau')
 
     def testNiceFormattingForEllipsis(self):
-        m = mock()
+        obj = mock()
+        m = mock_registry.mock_for(obj)
```
These were actually correct.  We use a fake object (aka `mock({"strict": False)`) as the first argument to `StubbedInvocation`.

kaste commented on tests/ellipsis_test.py:
```diff
@@ -373,23 +374,26 @@ def testKwargsMustBeUsedAsStarKwarg(self):
 
         assert repr(inv) == 'bark(...)'
 
     def testNiceFormattingForArgs(self):
-        m = mock()
+        obj = mock()
+        m = mock_registry.mock_for(obj)
```
Revert as above

kaste commented on tests/ellipsis_test.py:
```diff
@@ -373,23 +374,26 @@ def testKwargsMustBeUsedAsStarKwarg(self):
         inv(*args)
 
         assert repr(inv) == 'bark(*args)'
 
     def testNiceFormattingForKwargs(self):
-        m = mock()
+        obj = mock()
+        m = mock_registry.mock_for(obj)
```
Revert as above

kaste commented on tests/ellipsis_test.py:
```diff
@@ -177,6 +177,7 @@ def testEllipsisAsSecondArgumentPasses(self, call):
 
         assert rex.bark(*call.args, **call.kwargs) == 'Miau'
 
+
```
Revert
```suggestion

```

kaste commented on mockito/mocking.py:
```diff
@@ -144,20 +140,24 @@ def new_mocked_method(*args, **kwargs):
             self.replace_method(method_name, original_method)
 
     def forget_stubbed_invocation(self, invocation):
+
```
```suggestion

```

kaste commented on mockito/invocation.py:
```diff
@@ -463,6 +506,5 @@ def answer(self, *args, **kwargs):
             a = self.answers[0]
         else:
             a = self.answers.popleft()
-
```
Revert WS change

kaste commented:
That's the default configuration/style from isort?   

I was an early adaptor of black, but they just never really evolved.  It just produces too many really bad, ill formatted lines.  (It should really produce **none** ill formatted lines to be clear because otherwise it becomes an annoyance fast.)   I also liked the obviously javascript informed style of black first, but doing a lot more python than before I actually find the python standard library style would be most of the time👍.

avandierast commented:
For isort, I use the last version with the black profile :D But the profile doesn't seem to change anything here.
So yes, I think that is the default configuration.

kaste commented:
❤️ it.  Great work.

avandierast commented:
Thanks ! Great review ;)

kaste commented:
Here we go.
